### PR TITLE
Extend gather_summary_obs with localization

### LINKED ISSUE #13666: gather_summary_observations does not handle existing localization values.

### PR DESCRIPTION
**Pull request by SAKavli**
This collects localization data for each summary key and produces a WELL { LOCALIZATION { ... } } in the printed output.

**Issue**
Resolves #13666 

**Approach**
_Short description of the approach_

(Screenshot of new behavior in GUI if applicable)

- [x] PR title captures the intent of the changes, and is fitting for release notes.
- [x] Added appropriate release note label
- [ ] Commit history is consistent and clean, in line with the [contribution guidelines](https://github.com/equinor/ert/blob/main/CONTRIBUTING.md).
- [ ] Make sure unit tests pass locally after every commit (`git rebase -i main
      --exec 'just rapid-tests'`)

## When applicable
- [ ] **When there are user facing changes**: Updated documentation
- [ ] **New behavior or changes to existing untested code**: Ensured that unit tests are added (See [Ground Rules](https://github.com/equinor/ert/blob/main/CONTRIBUTING.md#ground-rules)).
- [ ] **Large PR**: Prepare changes in small commits for more convenient review
- [ ] **Bug fix**: Add regression test for the bug
- [ ] **Bug fix**: Add backport label to latest release (format: 'backport release-branch-name')

### CODE REVIEW

File: src/ert/gather_summary_observations.py [hunk 1]
```diff
@@ -152,11 +157,52 @@ def non_empty_fields(skds: list[SummaryKeyData]) -> list[str]:
...
+    if east is None or north is None:
+        return None
+    lines = [
+        f"{INDENT4}LOCALIZATION {{",
+        f"{INDENT6}EAST={east};",
+        f"{INDENT6}NORTH={north};",
+    ]
+    if radius is not None:
+        lines.append(f"{INDENT6}RADIUS={radius};")
+    lines.append(f"{INDENT4}}};")
```
Comment (SAKavli) [inline src/ert/gather_summary_observations.py:172]: Double }} escapes the f string, resulting in a single }

**Comment by SAKavli** [comment]
Copilot have made valid points, let me fix this.

File: tests/ert/ui_tests/cli/test_gather_summary_observations.py [hunk 2]
```diff
@@ -90,3 +91,70 @@ def test_that_single_experiment_in_storage_is_automatically_selected(capsys):
         "WOPR, OP1, 0.3, 0.075, 2012-12-15",
     ]
     assert all(line in csv_content for line in expected_csv_content)
+
+
+@pytest.mark.usefixtures("use_tmpdir")
+def test_that_convert_summary_observations_extracts_localization_information(capsys):
```
Comment (achaikou) [inline tests/ert/ui_tests/cli/test_gather_summary_observations.py]: Why is this test located in `ui_tests` ? Isn't that a unit_test because it doesn't actually call the cli?

File: src/ert/gather_summary_observations.py [hunk 3]
```diff
@@ -152,11 +157,52 @@ def non_empty_fields(skds: list[SummaryKeyData]) -> list[str]:
...
+    return "\n".join(lines)
+
+
+def get_first_loc_value(loc_key: str, summary_obs: list[Any]) -> float | None:
```
Comment (achaikou) [inline src/ert/gather_summary_observations.py:176]: If find this function very hard to read.

Unreadable `next` is one thing, but this confuses me a lot:
`for val in obs.get(loc_key, [None])`
There shouldn't be for example two "east" values for observation?
So why is the observation format like this - a list? From what I recall it is usually one nice value.

Anyway I think something straightforward would be more readable
```
def get_first_loc_value(loc_key: str, summary_obs: list[Any]) -> float | None:
    for obs in summary_obs:
        key_value = obs.get(loc_key, [None])[0]
        if key_value is not None:
            return float(key_value)
    return None
```

File: src/ert/gather_summary_observations.py [hunk 4]
```diff
@@ -152,11 +157,52 @@ def non_empty_fields(skds: list[SummaryKeyData]) -> list[str]:
...
+    for key in summary_observations:  # noqa: PLC0206
+        summary_key = make_summary_key_data(key)
+        skds.append(summary_key)
+        # Pick the first localization value which is not None for all localization
+        # keywords. These might get fetched from different observations.
+        east = get_first_loc_value("east", summary_observations[key])
+        north = get_first_loc_value("north", summary_observations[key])
+        radius = get_first_loc_value("radius", summary_observations[key])
+        loc_string = localization_to_string(east, north, radius)
+        if loc_string is not None and (well := summary_key.well) is not None:
```
Comment (achaikou) [inline src/ert/gather_summary_observations.py:197]: Might be nice adding to your test an entry starting not with the "well" keyword, but some "xxx" and having east-north values.
Free test path! 💰

**Review by eivindjahren — APPROVED** [review]

File: src/ert/gather_summary_observations.py [hunk 5]
```diff
@@ -152,11 +157,52 @@ def non_empty_fields(skds: list[SummaryKeyData]) -> list[str]:
...
-    print(f"SUMMARY {{\n  VALUES = {csv_file_name};\n}};\n")
-    skds = [make_summary_key_data(key) for key in summary_observations]
+    skds = []
+
+    localization = {}
+    for key in summary_observations:  # noqa: PLC0206
+        summary_key = make_summary_key_data(key)
+        skds.append(summary_key)
+        # Pick the first localization value which is not None for all localization
+        # keywords. These might get fetched from different observations.
```
Comment (achaikou) [inline src/ert/gather_summary_observations.py]: Text is confusing. I am thinking that we are getting the first value that is not None for all localization keywords together. Yet your second sentence contradicts this.
Maybe change the order and all -> each.

> For each localization keyword, pick the first localization value which is not None.

Also I am very confused why this behavior is acceptable. What are we trying to gather and print here, if east could come from OBS1 and north from OBS10 and radius from OBS30?

Or do we assume that all "well" observations should for some reason have reasonably similar east-north values?

**Review by achaikou — APPROVED** [review]
Nice! 🐱 
Just a few minor comments from me

File: tests/ert/ui_tests/cli/test_gather_summary_observations.py (hunk 2 above)
Comment (SAKavli) [reply tests/ert/ui_tests/cli/test_gather_summary_observations.py]: This is true! I can move it =)

File: src/ert/gather_summary_observations.py (hunk 3 above)
Comment (SAKavli) [reply src/ert/gather_summary_observations.py:176]: Thanks for letting me know!
You are probably right, but I just didn't make the assumption that it is always found on first index, but perhaps this holds. I guess I used next() to safeguard if there is some edge case I haven't thought of, but I can go with your suggestion!

File: src/ert/gather_summary_observations.py (hunk 5 above)
Comment (SAKavli) [reply src/ert/gather_summary_observations.py]: The purpose of this code is to print a bulk config in terminal, looking like this:

```
    SUMMARY {
      VALUES = foo.csv;
      WELL OP1 {
        LOCALIZATION {
          EAST=10;
          NORTH=20;
          RADIUS=2500;
        };
      };
    };
```
In this format, we assume that localization is the same for all observations from the same well.
Therefore, I make the simplification that we pick the first localization values as we have no better guess on which one to use.
I suppose we could raise an error for this, but for prototyping purposes I think we can revisit this once we land on the design and purpose.

But I agree, the comment is confusing, I can change it.

Comment (SAKavli) [reply src/ert/gather_summary_observations.py]: I will just remove it actually, it is a little confusing since the localization value for the same well will actually be overwritten by the last iteration of the same well in the for loop. I can perhaps write an issue to document this confusing behavior which may be revisited.

File: src/ert/gather_summary_observations.py (hunk 4 above)
Comment (SAKavli) [reply src/ert/gather_summary_observations.py:197]: I am not sure if I followed why it shouldn't start with the well keyword?
There is one entry already `WELL_WITHOUT_RADIUS` - is this what you mean?

Comment (achaikou) [reply src/ert/gather_summary_observations.py:197]: I mean that you could add an entry to the tests where well is None, what you check in this line.
From what I understood name should start with W "property" to be considered a well.

So entry like
```

        "XXX:UNWELL": [
            {
                "name": "XXX",
                "errors": [0.05],
                "values": [0.1],
                "east": [10],
                "north": [20],
                "radius": [2500],
                "x_axis": ["2010-03-31T00:00:00"],
            },
        ],
```

will not affect test results, but would add a test for non-well entries.
